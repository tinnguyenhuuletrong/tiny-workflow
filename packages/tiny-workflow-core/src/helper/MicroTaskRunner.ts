export class MicroTaskRunner {
  private pendingTasks: Promise<any>[] = [];
  private resolved: PromiseSettledResult<any>[] = [];

  public addTask(
    task: Promise<any>,
    opts?: {
      runId: string;
    }
  ) {
    this.pendingTasks.push(task);
    return task;
  }

  reset() {
    this.pendingTasks.length = 0;
    this.resolved.length = 0;
  }

  stats() {
    return {
      total: this.pendingTasks.length,
      done: this.resolved.length,
    };
  }

  /** waiting for all pending tasks settled */
  async idle() {
    while (this.resolved.length != this.pendingTasks.length) {
      this.resolved = await Promise.allSettled(this.pendingTasks);
    }
  }
}
