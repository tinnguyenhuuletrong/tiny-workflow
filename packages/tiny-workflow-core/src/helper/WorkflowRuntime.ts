import { DurableState } from "../DurableState";
import type {
  Constructor,
  DurableStateSystemEntry,
  SnapshotType,
} from "../type";

export type WorkflowRuntimeOpt<T extends DurableState> = {
  ctx: IRuntimeContext<T>;
  genRunId: () => Promise<string>;
  InstanceClass: Constructor<T>;
};

export interface IRuntimeContext<T extends DurableState> {
  attachRuntime(runtime: WorkflowRuntime<T>): void;

  save(runId: string, snapshotData: SnapshotType<T>): Promise<void>;
  load(runId: string): Promise<SnapshotType<T> | null>;

  lock(runId: string): Promise<void>;
  unlock(runId: string): Promise<void>;
}

export type WorkflowRunResult =
  | {
      runId: string;
      status: "need_resume";
      resumeEntry: DurableStateSystemEntry;
    }
  | {
      runId: string;
      status: "finished";
    }
  | {
      runId: string;
      status: "error";
      error?: any;
    };

export class WorkflowRuntime<T extends DurableState> {
  constructor(private opts: WorkflowRuntimeOpt<T>) {
    this.opts.ctx.attachRuntime(this);
  }

  get ctx() {
    return this.opts.ctx;
  }

  async run(ins: T, runId?: string): Promise<WorkflowRunResult> {
    if (!runId) runId = await this.opts.genRunId();

    try {
      await this.ctx.lock(runId);

      for await (const it of ins.exec(runId)) {
        if (it.resumeTrigger) {
          const resumeId = it.resumeTrigger.resumeId;
          const resumeEntry = ins.getResume(resumeId);
          if (!resumeEntry)
            throw new Error(`ResumeEntry not found resumeId=${resumeId}`);

          const data = ins.toJSON() as SnapshotType<T>;
          await this.opts.ctx.save(runId, data);

          return {
            runId,
            status: "need_resume",
            resumeEntry,
          };
        }
      }

      const data = ins.toJSON() as SnapshotType<T>;
      await this.opts.ctx.save(runId, data);
      return { runId, status: "finished" };
    } catch (error) {
      return { runId, status: "error", error };
    } finally {
      await this.ctx.unlock(runId);
    }
  }

  async resume(
    runId: string,
    {
      resumeId,
      resumePayload,
    }: {
      resumeId: string;
      resumePayload?: any;
    }
  ) {
    const ins = await this.createInstanceByRunId(
      runId,
      this.opts.InstanceClass
    );
    ins.resolveResume(resumeId, resumePayload);
    return this.run(ins, runId);
  }

  async createInstanceByRunId(runId: string, Cons: Constructor<T>) {
    const data = await this.opts.ctx.load(runId);
    if (!data) throw new Error(`snapshotData not found for runId=${runId}`);
    const ins = DurableState.fromJSON(Cons, data) as T;
    return ins;
  }
}
