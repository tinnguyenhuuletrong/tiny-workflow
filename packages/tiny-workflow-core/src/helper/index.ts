export * from "./MicroTaskRunner";
export * from "./SimpleContext";
export * from "./WorkflowRuntime";
