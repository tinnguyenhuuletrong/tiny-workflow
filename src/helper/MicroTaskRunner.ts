export class MicroTaskRunner {
  private pendingTasks: Promise<any>[] = [];
  private resolved: PromiseSettledResult<any>[] = [];

  public addTask(task: Promise<any>) {
    this.pendingTasks.push(task);
  }

  reset() {
    this.pendingTasks.length = 0;
    this.resolved.length = 0;
  }

  stats() {
    return {
      total: this.pendingTasks.length,
      done: this.resolved.length,
    };
  }

  async idle() {
    while (this.resolved.length != this.pendingTasks.length) {
      this.resolved = await Promise.allSettled(this.pendingTasks);
    }
  }
}
