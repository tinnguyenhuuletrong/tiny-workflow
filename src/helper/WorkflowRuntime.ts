import { DurableState } from "../DurableState";
import type { Constructor, DurableStateSystemEntry } from "../type";

export type SnapshotType<T extends DurableState> = ReturnType<T["toJSON"]>;
export type WorkflowRuntimeOpt<T extends DurableState> = {
  storage: IStorage<T>;
  genRunId: () => Promise<string>;
};

export interface IStorage<T extends DurableState> {
  save(runId: string, snapshotData: SnapshotType<T>): Promise<void>;
  load(runId: string): Promise<SnapshotType<T> | null>;
}

type WorkflowRunResult =
  | {
      runId: string;
      status: "need_resume";
      resumeEntry: DurableStateSystemEntry;
    }
  | {
      runId: string;
      status: "finished";
    };

export class WorkflowRuntime<T extends DurableState> {
  constructor(private opts: WorkflowRuntimeOpt<T>) {}

  async run(ins: T, runId?: string): Promise<WorkflowRunResult> {
    if (!runId) runId = await this.opts.genRunId();

    for await (const it of ins.exec(runId)) {
      if (it.resumeTrigger) {
        const resumeId = it.resumeTrigger.resumeId;
        const resumeEntry = ins.getResume(resumeId);
        if (!resumeEntry)
          throw new Error(`ResumeEntry not found resumeId=${resumeId}`);

        const data = ins.toJSON() as SnapshotType<T>;
        await this.opts.storage.save(runId, data);

        return {
          runId,
          status: "need_resume",
          resumeEntry,
        };
      }
    }

    const data = ins.toJSON() as SnapshotType<T>;
    await this.opts.storage.save(runId, data);
    return { runId, status: "finished" };
  }

  async resume(
    runId: string,
    Cons: Constructor<T>,
    {
      resumeId,
      resumePayload,
    }: {
      resumeId: string;
      resumePayload?: any;
    }
  ) {
    const ins = await this.getInsByRunId(runId, Cons);
    ins.resolveResume(resumeId, resumePayload);
    return this.run(ins, runId);
  }

  async getInsByRunId(runId: string, Cons: Constructor<T>) {
    const data = await this.opts.storage.load(runId);
    if (!data) throw new Error(`snapshotData not found for runId=${runId}`);
    const ins = DurableState.fromJSON(Cons, data) as T;
    return ins;
  }
}
