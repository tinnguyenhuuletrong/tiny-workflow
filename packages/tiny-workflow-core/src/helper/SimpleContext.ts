import { setTimeout as waitMs } from "node:timers/promises";
import type { DurableState } from "../DurableState";
import type { SnapshotType } from "../type";
import { MicroTaskRunner } from "./MicroTaskRunner";
import type {
  IRuntimeContext,
  WorkflowRuntime,
  WorkflowRunResult,
} from "./WorkflowRuntime";

export class MemoryLock {
  readonly promise: Promise<void>;
  private resolve!: Function;

  constructor() {
    this.promise = new Promise<void>((resolve) => {
      this.resolve = resolve;
    });
  }

  unlock() {
    this.resolve();
  }

  async wait() {
    return this.promise;
  }
}

export class SimpleContext<T extends DurableState>
  implements IRuntimeContext<T>
{
  readonly runner = new MicroTaskRunner();
  runtime!: WorkflowRuntime<T>;

  db = new Map<string, SnapshotType<T>>();
  dbLock = new Map<string, MemoryLock>();

  attachRuntime(runtime: WorkflowRuntime<T>) {
    if (this.runtime)
      throw new Error("can not re-attach. Something wrong here");
    this.runtime = runtime;
  }

  async lock(runId: string) {
    const tmp = this.dbLock.get(runId);
    if (tmp) await tmp.wait();

    const ins = new MemoryLock();
    this.dbLock.set(runId, ins);
  }

  async unlock(runId: string) {
    const tmp = this.dbLock.get(runId);
    if (tmp) {
      tmp.unlock();
      this.dbLock.delete(runId);
    }
  }

  async save(runId: string, snapshotData: SnapshotType<T>) {
    this.db.set(runId, snapshotData);
  }

  async load(runId: string) {
    const tmp = this.db.get(runId);
    if (!tmp) return null;
    return tmp;
  }

  async start() {
    this.runner.reset();
  }

  async shutdown() {
    for (const it of this.dbLock.entries()) {
      it[1].unlock();
    }
  }

  async withTransaction(runId: string, action: () => Promise<void>) {
    try {
      await this.lock(runId);
      await action();
    } catch (error) {
      throw error;
    } finally {
      await this.unlock(runId);
    }
  }

  async scheduleNextRun(runId: string, result: WorkflowRunResult) {
    switch (result.status) {
      case "need_resume":
        {
          switch (result.resumeEntry.type) {
            case "timer":
              {
                await waitMs(result.resumeEntry.resumeAfter - Date.now());
                const res = await this.runtime.resume(runId, {
                  resumeId: result.resumeEntry.resumeId,
                  resumePayload: undefined,
                });

                this.runner.addTask(this.scheduleNextRun(runId, res), {
                  runId,
                });
              }
              break;

            default:
              throw new Error(
                `can not handle resumeType=${result.resumeEntry.type}`
              );
          }
        }
        break;

      case "error": {
        console.error(`error runId=${runId} - detail=`, result);
        return;
      }
      default:
        break;
    }
  }
}
