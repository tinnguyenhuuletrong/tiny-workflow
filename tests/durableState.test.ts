import { test, expect, mock } from "bun:test";
import {
  DurableState,
  type ContinueTrigger,
  type DurableStateIterator,
  type StepHandler,
  type StepIt,
} from "../src";
import * as TestHelper from "./helper";

test("simple_durable_state", async () => {
  enum EStep {
    step_begin = "step_begin",
    step_compute = "step_compute",
    step_end = "step_end",
  }
  type TStateShape = Partial<{
    startHit: boolean;
    computed: number;
    endHit: boolean;
  }>;
  type EAuditLog = "start_hit" | "end_hit" | "computed_hit";

  class SampleState extends DurableState<EStep, TStateShape, EAuditLog> {
    // Type Safeguard & Auto register
    private _static(key: EStep): StepHandler<EStep> {
      return this[key];
    }
    private _collectAndRegisterSteps() {
      Object.values(EStep).map((step) =>
        this.stepHandler.set(step, this._static(step).bind(this))
      );
    }
    constructor() {
      super(EStep.step_begin, {
        withAuditLog: true,
      });
      this._collectAndRegisterSteps();
    }

    private async *step_begin(): StepIt<EStep> {
      this.state.startHit = true;
      this.addLog({
        type: "start_hit",
        values: {},
      });
      return { nextStep: EStep.step_compute };
    }

    private async *step_compute(): StepIt<EStep> {
      // with cached action
      const { it, value } = await this.withAction(
        "cache_computed",
        async () => {
          return Math.PI;
        }
      );
      if (it) yield it;

      this.addLog({
        type: "computed_hit",
        values: {
          actionValue: value,
        },
      });
      this.state.computed = value;

      return { nextStep: EStep.step_end };
    }
    private async *step_end(): StepIt<EStep> {
      this.state.endHit = true;
      this.addLog({
        type: "end_hit",
        values: {},
      });
      return { nextStep: null };
    }
  }

  // Test
  const ins = DurableState.fromJSON<EStep, TStateShape, EAuditLog, SampleState>(
    SampleState,
    {
      step: EStep.step_begin,
      state: {
        computed: 0,
      },
    }
  );
  const it = ins.exec();
  for await (const v of it) {
  }
  expect(ins.currentState).toMatchSnapshot("final_state");
  expect(TestHelper.removeProps(ins.auditLogs, ["_at"])).toMatchSnapshot(
    "audit_logs"
  );
});

test("step_durable_state", async () => {
  enum EStep {
    step_begin = "step_begin",
    step_compute = "step_compute",
    step_end = "step_end",
  }
  type TStateShape = Partial<{
    startHit: boolean;
    computed: number;
    endHit: boolean;
  }>;
  type EAuditLog = "start_hit" | "end_hit" | "computed_hit";

  class SampleState extends DurableState<EStep, TStateShape, EAuditLog> {
    // Type Safeguard & Auto register
    private _static(key: EStep): StepHandler<EStep> {
      return this[key];
    }
    private _collectAndRegisterSteps() {
      Object.values(EStep).map((step) =>
        this.stepHandler.set(step, this._static(step).bind(this))
      );
    }
    constructor() {
      super(EStep.step_begin, {
        withAuditLog: true,
      });
      this._collectAndRegisterSteps();
    }

    private async *step_begin(): StepIt<EStep> {
      this.state.startHit = true;
      this.addLog({
        type: "start_hit",
        values: {},
      });
      return { nextStep: EStep.step_compute };
    }

    private async *step_compute(): StepIt<EStep> {
      let sum = 0;
      for (let i = 0; i < 3; i++) {
        // with cached action
        const { it, value } = await this.withAction(
          `cache_computed_${i}`,
          async () => {
            return Math.PI;
          }
        );
        if (it) yield it;
        sum += value;
      }

      this.addLog({
        type: "computed_hit",
        values: {
          actionValue: sum,
        },
      });
      this.state.computed = sum;

      return { nextStep: EStep.step_end };
    }
    private async *step_end(): StepIt<EStep> {
      this.state.endHit = true;
      this.addLog({
        type: "end_hit",
        values: {},
      });
      return { nextStep: null };
    }
  }

  // Test
  const ins = DurableState.fromJSON<EStep, TStateShape, EAuditLog, SampleState>(
    SampleState,
    {
      step: EStep.step_begin,
      state: {
        computed: 0,
      },
    }
  );

  // 1st run
  const it = ins.exec();
  let res = await it.next();
  expect(res.done).toBe(false);
  let snapshot = TestHelper.removeProps(ins.toJSON(), ["_at"]);
  expect(snapshot).toMatchSnapshot("run_part_1");

  // 2nd run
  res = await it.next();
  res = await it.next();
  res = await it.next();
  expect(res.done).toBe(true);
  snapshot = TestHelper.removeProps(ins.toJSON(), ["_at"]);
  expect(snapshot).toMatchSnapshot("run_part_2");
});

test("wait_ms_durable_state", async () => {
  enum EStep {
    step_begin = "step_begin",
    step_end = "step_end",
  }
  type TStateShape = Partial<{
    startHit: boolean;
    endHit: boolean;
  }>;
  type EAuditLog = "log";

  class SampleState extends DurableState<EStep, TStateShape, EAuditLog> {
    // Type Safeguard & Auto register
    private _static(key: EStep): StepHandler<EStep> {
      return this[key];
    }
    private _collectAndRegisterSteps() {
      Object.values(EStep).map((step) =>
        this.stepHandler.set(step, this._static(step).bind(this))
      );
    }
    constructor() {
      super(EStep.step_begin, {
        withAuditLog: true,
      });
      this._collectAndRegisterSteps();
    }

    private async *step_begin(): StepIt<EStep> {
      this.state.startHit = true;
      const { it } = this.waitForMs("wait_for_1sec", 1000);
      if (it) yield it;
      this.addLog({
        type: "log",
        values: {
          msg: "after 1 sec",
        },
      });

      return { nextStep: EStep.step_end };
    }

    private async *step_end(): StepIt<EStep> {
      this.state.endHit = true;

      return { nextStep: null };
    }
  }

  // Test
  const ins = DurableState.fromJSON<EStep, TStateShape, EAuditLog, SampleState>(
    SampleState,
    {
      step: EStep.step_begin,
      state: {},
    }
  );

  (ins as any).genResumeId = (key: string) => `${key}-1`;
  const it = ins.exec();
  let res = await it.next();

  // not finish - have a continue trigger
  expect(res.done).toBe(false);
  const resume = (res.value as DurableStateIterator<EStep>)
    .resumeTrigger as ContinueTrigger;
  expect(resume).not.toBeNull();
  expect(TestHelper.removeProps(resume, ["resumeAt"])).toMatchSnapshot();

  // when time on
  ins.resolveResume(resume.resumeId, {});

  res = await it.next();
  expect(res.done).toBe(true);

  // snapshot check
  let snapshot = TestHelper.removeProps(ins.toJSON(), [
    "_at",
    "resumeAt",
    "resumeAfter",
  ]);
  expect(snapshot).toMatchSnapshot("durable_state_with_waiting");
});
