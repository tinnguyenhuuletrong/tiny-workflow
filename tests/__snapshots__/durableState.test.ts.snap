// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`simple_durable_state: final_state 1`] = `
{
  "computed": 3.141592653589793,
  "endHit": true,
  "startHit": true,
}
`;

exports[`simple_durable_state: audit_logs 1`] = `
[
  {
    "_step": "step_begin",
    "type": "start_hit",
    "values": {},
  },
  {
    "_step": "step_begin",
    "type": "transition",
    "values": {
      "from": "step_begin",
      "to": "step_compute",
    },
  },
  {
    "_step": "step_compute",
    "type": "cache",
    "values": {
      "key": "step_compute:cache_computed",
    },
  },
  {
    "_step": "step_compute",
    "type": "computed_hit",
    "values": {
      "actionValue": 3.141592653589793,
    },
  },
  {
    "_step": "step_compute",
    "type": "transition",
    "values": {
      "from": "step_compute",
      "to": "step_end",
    },
  },
  {
    "_step": "step_end",
    "type": "end_hit",
    "values": {},
  },
]
`;

exports[`step_durable_state: run_part_1 1`] = `
{
  "cache": {
    "step_compute:cache_computed_0": 3.141592653589793,
  },
  "logs": [
    {
      "_step": "step_begin",
      "type": "start_hit",
      "values": {},
    },
    {
      "_step": "step_begin",
      "type": "transition",
      "values": {
        "from": "step_begin",
        "to": "step_compute",
      },
    },
    {
      "_step": "step_compute",
      "type": "cache",
      "values": {
        "key": "step_compute:cache_computed_0",
      },
    },
  ],
  "state": {
    "computed": 0,
    "startHit": true,
  },
  "step": "step_compute",
  "system": {},
}
`;

exports[`step_durable_state: run_part_2 1`] = `
{
  "cache": {
    "step_compute:cache_computed_0": 3.141592653589793,
    "step_compute:cache_computed_1": 3.141592653589793,
    "step_compute:cache_computed_2": 3.141592653589793,
  },
  "logs": [
    {
      "_step": "step_begin",
      "type": "start_hit",
      "values": {},
    },
    {
      "_step": "step_begin",
      "type": "transition",
      "values": {
        "from": "step_begin",
        "to": "step_compute",
      },
    },
    {
      "_step": "step_compute",
      "type": "cache",
      "values": {
        "key": "step_compute:cache_computed_0",
      },
    },
    {
      "_step": "step_compute",
      "type": "cache",
      "values": {
        "key": "step_compute:cache_computed_1",
      },
    },
    {
      "_step": "step_compute",
      "type": "cache",
      "values": {
        "key": "step_compute:cache_computed_2",
      },
    },
    {
      "_step": "step_compute",
      "type": "computed_hit",
      "values": {
        "actionValue": 9.42477796076938,
      },
    },
    {
      "_step": "step_compute",
      "type": "transition",
      "values": {
        "from": "step_compute",
        "to": "step_end",
      },
    },
    {
      "_step": "step_end",
      "type": "end_hit",
      "values": {},
    },
  ],
  "state": {
    "computed": 9.42477796076938,
    "endHit": true,
    "startHit": true,
  },
  "step": "step_end",
  "system": {},
}
`;

exports[`wait_ms_durable_state 1`] = `
{
  "resumeId": "wait_for_1sec-1",
  "type": "time",
}
`;

exports[`wait_ms_durable_state: durable_state_with_waiting 1`] = `
{
  "cache": {},
  "logs": [
    {
      "_step": "step_begin",
      "type": "interrupt_begin",
      "values": {
        "resumeId": "wait_for_1sec-1",
        "type": "time",
      },
    },
    {
      "_step": "step_begin",
      "type": "interrupt_end",
      "values": {
        "resumeId": "wait_for_1sec-1",
      },
    },
    {
      "_step": "step_begin",
      "type": "log",
      "values": {
        "msg": "after 1 sec",
      },
    },
    {
      "_step": "step_begin",
      "type": "transition",
      "values": {
        "from": "step_begin",
        "to": "step_end",
      },
    },
  ],
  "state": {
    "endHit": true,
    "startHit": true,
  },
  "step": "step_end",
  "system": {
    "step_begin:timer:wait_for_1sec": {
      "isDone": true,
      "responsePayload": {},
      "resumeId": "wait_for_1sec-1",
      "type": "timer",
    },
  },
}
`;
